lunabot_control/scripts/homing_controller.py

```diff
@@ -0,0 +1,332 @@
+
+    def _robot_state_cb(self, msg):
+        self._drive_left_curr = msg.drive_left_curr
+        self._drive_right_curr = msg.drive_right_curr
+        self._dep_curr = msg.dep_curr
+
+    def homing(self):
+        if self.T_camera_april_msg is None:
+            return
+        T_camera_april = ros_numpy.numpify(
+            self.T_camera_april_msg
+        )  # Converting the transformation matrix to a numpy array
+
+        print(T_camera_april)
+
+        # Assigning the unit vector values for the different vectors
+        v_1A = np.array([0, 0, -1, 1])
+        # V_1A refers to the april tag reference in the april tag frame
+        v_2C = np.array([1, 0, 0, 1])
+        # V_2C refers to the unit vector for the camera frame of the robot
+
+        # compare = np.array([0,0,0,0])
+
+        # Computing the location of point V that is in frame A to transformed to its location in frame C
+        v_1C = T_camera_april @ v_1A
+        p_A_zero = np.array([0, 0, 0, 1])
+        v_C_april = T_camera_april @ p_A_zero
+        # if (T_camera_april == compare):
+        #     self._effort_msg.left_drive = 0
+        #     self._effort_msg.right_drive = 0
+
+        v_1C[2] = 0
+        v_C_april[2] = 0
+        v_C_april[3] = 0
+
+        # Finding the error angle between the two points, v_1C and v_2C, using dot product (in radians)
+
+        # v_1C = v_1C[0:2]
+        # v_2C = v_2C[0:2]
+
+        angle = np.arctan2(
+            v_1C[0] * v_2C[1] - v_2C[0] * v_1C[1], v_1C[0] * v_1C[1] + v_2C[0] * v_2C[1]
+        )
+
+        angle += 3 * np.pi / 2 + 0.3
+        if angle > np.pi:
+            angle -= 2 * np.pi
+
+        # determining the linear translational error (in x and y) of the robot considered
+
+        self.v_C_april = np.array(
+            [
+                self.T_camera_april_msg.position.x,
+                self.T_camera_april_msg.position.z,
+                0,
+                0,
+            ]
+        )
+
+        print(v_C_april)
+        distance = np.linalg.norm(v_C_april)
+        # Replaced v_2c with 0,0,0,1 for the sim because the distance being calculated
+        # Was seemingly off from the apriltag (based on mapping out values, was 0,0,0)
+
+        # write PD controller here with the output being lin, ang velocity
+        # Define the K constants for P, I, and D
+        # For sim tuning, lin and ang are flipped.
+        # For Testing angular tuning only change the first value
+
+        # Define the errors
+        # lin_error_dist = np.sqrt((lin_error[0])**2 + (lin_error[1])**2)
+        self._curr_error = np.array(
+            [distance - self.linear_setpoint, angle - self.angular_setpoint]
+        )
+
+        print("self._curr_error: ", self._curr_error)
+        self._error_total += self._curr_error
+
+        # Computing PID errors
+        ctrl = self._curr_error * self.KP
+        ctrl += self._error_total * self.KI
+        ctrl += (self._curr_error - self._prev_error) * self.KD
```

Comment:
There's no d(delta) t in these calculations. When you integrate (KI) multiply by dt and when you differentiate (KD) divide by dt

Reply:
Would explain why d had little impact, its magnitude was at minimum 1/10 (if going at 10hz) of what it should be

Reply:
Shouldn't the tuning gains take care of that though? or will the values we're using still be an order of magnitude off?

Reply:
In theory yeah but better to use it. And the dt can change between each cycle right? Better to calc it using current and prev time rather than assuming a consistent hz

Reply:
Side note if your D is that hard to tune maybe reduce P a bit

Reply:
will try!

Reply:
Actually you're looking at the homing controller code, which we tested already and got working last time.